Kajabi/sage-lib
# Modal transition & animation settings

## Description
This branch adds a default fade transition to all modals along with options for animated movement. Screens attached below, though the effect should be easier to see on-page.

Transitions are controlled through `data` attributes. Using this implementation as a test case for an extended Sage animation library. 

Notes:
- huge thanks to @philschanely for assistance with the React version 🙇🏼 
- added ability to disable background blur: in testing, the css `filter` tends to increase GPU/CPU usage when animating, resulting in dropped frames


### Screenshots

|  Before  |  After  |
|--------|--------|
|![modal-before](https://user-images.githubusercontent.com/816579/110717750-7d3f4580-81be-11eb-808d-20ae74f6262a.gif)|![modal-after](https://user-images.githubusercontent.com/816579/110717712-6dbffc80-81be-11eb-81f3-f56916e48970.gif)|

## Test notes
In order to implement these transitions, the inactive modal CSS was changed from `display: none` to instead use `visibility: hidden` and a negative `z-index`. This should still be sufficient to "hide" the modal from accidental user interaction, but we'll need to ensure the stacking order is working as expected depending on placement of the modal's markup. In Rails, if `SageModal` is nested within a `<% content_for :modal do %>`, there should not be an issue.

All transitions are disabled when "reduce motion" is enabled on a user's OS, or on browsers that support the feature. Modal backdrop opacity is also increased for additional contrast.


### Steps to test
- Navigate to http://localhost:4000/pages/object/modal
- Open some modals
- (optional) Enable "reduce motion" in your MacOS system preferences -> Accessibility (screen below). All transitions should now be disabled

<img width="380" alt="" src="https://user-images.githubusercontent.com/816579/110720496-8383f080-81c3-11eb-80d7-5ceba7d98862.png">


### Estimated impact
- (HIGH) Modal CSS display property has been altered to allow for transitioning effects. Confirm that modal is displayed on page as expected, and cannot be accessed via user interaction or keyboard navigation. Examples of use:
  - [x] Beta features modal 
  - [x] Offer price modal
  - [x] Create site modal
  - [ ] Coupons: delete Coupon modal
  - [ ] Coupons: add Offer modal


## Related
